File: Year-2023/LeetCode-March-Challenge/DesignBrowserHistory.js
var BrowserHistory = function (homepage) {
    this.history = [homepage];
    this.curr = 0;
};


BrowserHistory.prototype.visit = function (url) {
    this.history.splice(this.curr + 1, this.history.length - this.curr + 1);
    this.history.push(url);
    this.curr++;
};


BrowserHistory.prototype.back = function (steps) {
    this.curr = Math.max(0, this.curr - steps);
    return this.history[this.curr];
};


BrowserHistory.prototype.forward = function (steps) {
    this.curr = Math.min(this.history.length - 1, this.curr + steps);
    return this.history[this.curr];
};

File: Year-2023/LeetCode-March-Challenge/DesignBrowserHistory.java
import java.util.*;

class BrowserHistory {
    private ArrayList<String> history = new ArrayList<>();
    private int curr = -1;

    public BrowserHistory(String homepage) {
        history.add(homepage);
        curr++;
    }

    public void visit(String url) {
        history.subList(curr + 1, history.size()).clear();
        history.add(url);
        curr++;
    }

    public String back(int steps) {
        curr = Math.max(0, curr - steps);
        return history.get(curr);
    }

    public String forward(int steps) {
        curr = Math.min((int)history.size() - 1, steps + curr);
        return history.get(curr);
    }
}

public class DesignBrowserHistory {
    public static void main(String[] args) {

    }
}